### PR TITLE
Update configure's default compilers list

### LINKED ISSUE #1021: configure can't find Intel icx compiler
[config.log](https://github.com/hypre-space/hypre/files/13536398/config.log)
This is not a real problem, because one could workaround by specifying CC=mpiicx. Since the classic icc compiler is removed from Intel oneAPI 2024.0, it would be better to support the new icx compiler.

### PR DESCRIPTION
**Pull request by victorapm**
This PR updates the default compilers list used in the configure script such that:
-  The newer `icx` has precedence over `icc`
-  `mpiicx`/`mpiicpx` have precedence over `mpiicc`/`mpiicpc`
- `--enable-debug` uses `-O0 -g -Wall` as CFLAGS

Closes #1021

### CODE REVIEW

**Comment by liruipeng** [comment]
Not sure about the changes. Please see my comments. Thanks!

File: src/config/hypre_macros_misc.m4 [hunk 1]
```diff
@@ -102,29 +102,29 @@ then
       gcc|mpigcc|mpicc)
         CFLAGS="-O2"
         if test "$hypre_using_openmp" = "yes" ; then
-          CFLAGS="$CFLAGS -fopenmp"
-          LDFLAGS="$LDFLAGS -fopenmp"
+          CFLAGS+=" -fopenmp"
```
Comment (liruipeng) [inline src/config/hypre_macros_misc.m4:105]: Why these changes?

File: src/config/configure.in [hunk 2]
```diff
@@ -1720,9 +1720,9 @@ then
    else
       if test "$hypre_using_openmp" = "yes"
       then
-         AC_CHECK_PROGS(CC, [mpxlc mpixlc_r mpixlc mpiicc mpiicx mpigcc mpicc mpipgcc mpipgicc])
+         AC_CHECK_PROGS(CC, [mpicc mpxlc mpixlc_r mpixlc mpiicc mpiicx mpigcc mpipgcc mpipgicc])
```
Comment (liruipeng) [inline src/config/configure.in]: This changes the behavior of hypre's configure at LLNL machines, such as lassen, where xlc is picked by default.

File: src/config/configure.in [hunk 3]
```diff
@@ -1738,16 +1738,16 @@ then
    then
       if test "$hypre_using_openmp" = "yes"
       then
-         AC_CHECK_PROGS(CXX, [xlC_r xlc_r xlC xlc icpc icc icpx icx g++ gcc pgCC pgcc pgc++ CC cc KCC kcc])
+         AC_CHECK_PROGS(CXX, [xlC_r xlc_r xlC xlc icx icpc icc icpx g++ gcc pgCC pgcc pgc++ CC cc KCC kcc])
```
Comment (liruipeng) [inline src/config/configure.in]: Even though icc is gone, icx can be picked  from the list. Not sure why we need to reorder.

File: src/config/hypre_macros_misc.m4 (hunk 1 above)
Comment (victorapm) [reply src/config/hypre_macros_misc.m4:105]: I think the new form is clearer and safer. Even though this example is simple, I think it's a good practice to use "+="

Comment (liruipeng) [reply src/config/hypre_macros_misc.m4:105]: Why is it safer?

File: src/config/configure.in (hunk 2 above)
Comment (victorapm) [reply src/config/configure.in]: `mpicc` is the same as `mpixlc` on Lassen. So this change seems fine to me:
``` 
vicmagri@lassen708 ~ $ mpixlc --version
IBM XL C/C++ for Linux, V16.1.1 (5725-C73, 5765-J13)
Version: 16.01.0001.0014
vicmagri@lassen708 ~ $ mpicc --version
IBM XL C/C++ for Linux, V16.1.1 (5725-C73, 5765-J13)
Version: 16.01.0001.0014
```

The main reason for this change is that `mpicc` is more general, so it makes sense to test for it first.

File: src/config/configure.in (hunk 3 above)
Comment (victorapm) [reply src/config/configure.in]: Some systems have a residual `icc` installation lying around. If we test for icx first, the residual `icc` wouldn't be a problem.

File: src/config/configure.in (hunk 2 above)
Comment (liruipeng) [reply src/config/configure.in]: They may not always be the same, under different module loads. If one wants to test mpicc, they can always do CC=mpicc. I think my question is still: why change given that this works fine for long time? :)

File: src/config/configure.in (hunk 3 above)
Comment (liruipeng) [reply src/config/configure.in]: Again, they can always do CC=WHATEVER, and after icc is complete gone, we just delete it from the list.

File: src/config/hypre_macros_misc.m4 (hunk 1 above)
Comment (victorapm) [reply src/config/hypre_macros_misc.m4:105]: Mainly because it reduces error risk since you don't need to type CFLAGS, for example, twice.

File: src/config/configure.in (hunk 3 above)
Comment (victorapm) [reply src/config/configure.in]: I agree users can always specify the compiler they want via CC, however the main aim here was to make the process of choosing compilers automatically (by the configure script) more robust. If such an improvement isn't necessary, I can close the PR :) @rfalgout thoughts?

Comment (liruipeng) [reply src/config/configure.in]: Hi @victorapm I don't really have that strong opinion on this PR :) My main concern if it is worth our time to change and test for some thing that is working OK. I am a "minimalist" :). Thanks!

**Comment by rfalgout** [comment]
Hi @victorapm and @liruipeng .  Both of you make good points, but let me make a general comment before replying individually to some of the changes.  The magic of configure or cmake or spack automatically finding compilers and such is nice to have, but it will not always work or work perfectly for every user.  So, we need to try to make it work for as many users as possible, while minimizing the impact on existing users.  This is a subjective call and I think most of your conversation has been about balancing these two conflicting goals.

File: src/config/hypre_macros_misc.m4 (hunk 1 above)
Comment (rfalgout) [reply src/config/hypre_macros_misc.m4:105]: I don't see any harm using the '+=' since we have moved away from using the Bourne Shell and have now adopted much of what is offered in bash.  I see the point about potentially making mistakes by typing the variable name in twice.  I also see the potential for mistakes by forgetting the extra space in the '+=' approach.  The '+=' syntax is more readable IMO.  I'm happy either way here.

File: src/config/configure.in (hunk 2 above)
Comment (rfalgout) [reply src/config/configure.in]: I think I agree with @liruipeng here that we should not reorder these.  This will change the behavior for many existing users.

File: src/config/configure.in (hunk 3 above)
Comment (rfalgout) [reply src/config/configure.in]: If we are reordering 'icc' (paired with 'icpc') and 'icx' (paired with 'icpx'), why are we not also reordering 'mpiicc' and 'mpiicx'?  I am not opposed to switching the order of these if 'icc' is going away, but shouldn't we do this consistently?  We tried to keep the mpi versions consistent with the raw compilers before.

File: src/config/hypre_macros_misc.m4 [hunk 4]
```diff
@@ -220,31 +220,31 @@ if test "x${hypre_user_chose_cflags}" = "xno"
 then
    case `basename ${CC}` in
       gcc|mpigcc|mpicc)
-        CFLAGS="-g -Wall"
+        CFLAGS="-O0 -g -Wall"
```
Comment (rfalgout) [inline src/config/hypre_macros_misc.m4:223]: I don't think we should add the '-O0' to '--enable-debug'.  I think this should just turn on debugging, not modify the optimization setting.  Why do you think this is a good thing to always do here?  Sometimes it is good to debug with different optimization settings (the bug may be caused by the optimization setting).

File: src/config/configure.in (hunk 3 above)
Comment (victorapm) [reply src/config/configure.in]: Good catch. `mpiicc` and `mpiicx` should be reordered as well. I can do that if we decide this PR is useful

File: src/config/hypre_macros_misc.m4 (hunk 4 above)
Comment (liruipeng) [reply src/config/hypre_macros_misc.m4:223]: Good catch. I didn't see this change since I thought they are all the "format changes". We also have `-O0` in CUFLAG with `--enable-debug`, which may not be a good thing either, see https://github.com/hypre-space/hypre/blob/70e9dec6bda36e6a60b98b58eff0656bb840d0d7/src/config/configure.in#L2543. The reason I added was optimization sometimes can remove information used by debuggers. I usually use debuggers like gdb with `--enable-debug`.

Comment (victorapm) [reply src/config/hypre_macros_misc.m4:223]: Because the compiler is free to choose the optimization level if not specified by the user. In many cases, the optimization level chosen by default is `-O2`. However, when debugging with gdb, `-O2` generally hides the value of local variables, which doesn't help debugging.

The point of a bug being related to a particular optimization flag is good. If that is the case, the developer trying to understand the bug could do `--with-extra-CFLAGS="-O2"`, for example, and `O2` would have the priority, thus bringing back the bug again

Comment (rfalgout) [reply src/config/hypre_macros_misc.m4:223]: Let's get a consensus on this question via teams.  For me, I use '--enable-debug' all the time and almost never have to additionally change the optimization flag (when I have done that in the past, it was typically because the optimization setting was causing the bug in the first place).  I usually use totalview, not gdb.  So, this will change the behavior of '--enable-debug' for me and I'm not sure if it will be for better or worse until I start using it.  My guess is I won't notice anything at all.

Comment (rfalgout) [reply src/config/hypre_macros_misc.m4:223]: Whatever we decide, we should be consistent.

File: src/config/configure.in (hunk 3 above)
Comment (rfalgout) [reply src/config/configure.in]: Okay, since the two compilers are right next to each other in the list, if we switch them, it should be "in place" and look as follows:
```
icpx icx icpc icc
mpiicx mpiicc
```
I don't know what the 'p' compilers are, but they were needed at some point so let's keep them there and in the order they are already given with respect to the non-'p' versions.

I am okay with making this change.  @liruipeng , what do you think?

I don't think we should move `mpicc` to a new position, however.  In general, the compiler orders here were originally intended to grab the most useful compiler on the LLNL systems first, so the general compilers tended to be near the end not the front.  But that was a long time ago and now it's more disruptive to change the orders of these things too much.

Comment (liruipeng) [reply src/config/configure.in]:
```
icpx icx icpc icc
mpiicx mpiicc
```
looks good to me.

File: src/config/configure.in (hunk 2 above)
Comment (victorapm) [reply src/config/configure.in]: Thanks for the comments! One issue with the current approach (specialized compilers first on the list) is that it doesn't guarantee that the most recent compiler loaded via `module load` will be the one picked by `configure`.

For example, on lassen, if we do `module load clang` and run `configure`, it will pick `CC=mpixlc` instead of `CC=mpiclang`. Moving `mpicc` to the top of the list is a solution to this problem (mpicc is effectively mpiclang in this scenario). Another solution, as we talked, is to explicitly set CC=mpiclang.

I don't see this as a big problem, because we have workarounds, but I just wanted to share it with you both...

File: src/config/hypre_macros_misc.m4 (hunk 4 above)
Comment (victorapm) [reply src/config/hypre_macros_misc.m4:223]: > the compiler is free to choose the optimization level [...] In many cases [...] default is -O2.

I did some research to clarify this and found the following defaults:
| Compiler | Opt. level |
|----------|---------------|
| `icx`/`icc` | `-O2` |
| `xlc` | `-O2` |
| `gcc` | `-O0` |
| `clang` | `-O0` |
| `msvc` | `-O0` |

I think it would be good to have an expected behavior with `--enable-debug`, either by ensuring `-O0` or `-O2`.

Also, `CMake` sets `-O0` with its `-DCMAKE_BUILD_TYPE=Debug` option. So, I think it makes sense to get the same with configure's `--enable-debug` for consistency.

**Comment by victorapm** [comment]
@rfalgout Does it make sense to remove insure (for the purpose of some cleaning up)? See:
https://github.com/hypre-space/hypre/blob/33524991d06f5e5b4c0c088ccf117aa09288867a/src/config/configure.in#L2106-L2124

File: src/config/configure.in (hunk 3 above)
Comment (victorapm) [reply src/config/configure.in]: Done in https://github.com/hypre-space/hypre/pull/1024/commits/91583f12640eeb455c152e153d632f4c58458d48

File: src/config/hypre_macros_misc.m4 (hunk 1 above)
Comment (victorapm) [reply src/config/hypre_macros_misc.m4:105]: Thanks! I'll consider this as resolved then :)

File: src/config/hypre_macros_misc.m4 (hunk 4 above)
Comment (rfalgout) [reply src/config/hypre_macros_misc.m4:223]: This seems reasonable to me.  @liruipeng , what do you think?

**Review by rfalgout — APPROVED** [review]
This looks good to me.  How does it affect the regression tests?

**Comment by rfalgout** [comment]
> @rfalgout Does it make sense to remove insure (for the purpose of some cleaning up)?

Yes, we could probably do that without upsetting anyone.  Thanks!

**Review by liruipeng — APPROVED** [review]
